### DISCUSSION TITLE
How to minibatch heterogeneous networks for link prediction?

### DISCUSSION THREAD

**Discussion by saiden89** [pyg-team/pytorch_geometric#6451] [Q&A]
Hi everyone,
I have a undirected heterogeneous graph with 1 type of node, but different kinds of edges, with the ultimate goal of performing link regression (as it is a probabilistic network with link weights between 0 and 1) on all edge types.

The current workflow first applies a `RandomLinkSplit` transformation to get the train, validation and test splits and then instantiates a `NeighborLoader` on each of the splits to get their respective loaders, which are then passed to a Pytorch Lightning model.
I can change the node loaders' parameters just fine in order to get different `batch_size`s, but would it be possible to batch the edges instead? I think this would pose some problems since the different edge types can differ for the number of elements even by orders of magnitude. Even so, should I then create different LinkLoaders for each edge type?
I'm not even sure that batching for edges even makes sense, and I would greatly appreciate your input on the matter!

**Comment by rusty1s** [comment]
I think what you want to do is use `LinkNeighborLoader` instead of `NeighborLoader` for this.

**Reply by saiden89** [reply]
I thought that much, but I'm confused about how this would work for edge types that have different number of elements. `LinkNeighboLoader` has the parameter `edge_label_index` that "*needs to be passed as a tuple that holds the edge type and corresponding edge indices.*". Given that I have several edge types in a heterogeneous graph, do I have to create different loaders for each edge type?
I have tried looking for examples of my use-case but there are none that cover both distribute learning with Lightning and heterogeneous link prediction.

**Reply by rusty1s** [reply]
Currently, we don't support sampling over different edge types. However, if all the edge types point to the same node type, You can consider converting your heterogeneous graph to a "typed homogeneous one" - that is a normal `Data` object and edge indices correspond to a given `edge_type` vector: `hetero_data.to_homogeneous()`.

**Reply by saiden89** [reply]
I have tried to do what you said, and indeed it work well until I have to convert back to heterodata when I also use`RandomLinkSplit`. It seems that no matter the order of link spitting and conversion, link type for the `edge_weight` gets lost. Example:
```python
from torch_geometric.datasets import FakeHeteroDataset
from torch_geometric.transforms import RandomLinkSplit

data = FakeHeteroDataset(num_graphs=1, num_node_types=1, num_edge_types=4, edge_dim=1, node_label=0)
transform = RandomLinkSplit(
    key="edge_weight",
    add_negative_train_samples=False,
    neg_sampling_ratio=0,
    edge_types=data[0].edge_types,
)

data = data[0].to_homogeneous(edge_attrs=["edge_weight"])
train, val, test = transform(data)
print(val)

>>> Data(node_label=[0], edge_index=[2, 22281], x=[802, 64], y=[802], node_type=[802], edge_type=[22281], edge_weight=[3183], edge_weight_index=[2, 3183])

```
For the supervision edges weight (`edge_weight`) there is no way of telling which weight belongs to which edge type. Moreover, when trying to convert back to heterogeneous
```python
train.to_heterogeneous()

---------------------------------------------------------------------------
IndexError                                Traceback (most recent call last)
Cell In [161], line 17
     15 train, val, test = transform(data)
     16 print(val)
---> 17 print(train.to_heterogeneous(edge_type=train.edge_type))
     18 # train.to_heterogeneous()

File lib/python3.10/site-packages/torch_geometric/data/data.py:683, in Data.to_heterogeneous(self, node_type, edge_type, node_type_names, edge_type_names)
    681         continue
    682     elif isinstance(value, Tensor) and self.is_node_attr(attr):
--> 683         data[key][attr] = value[node_ids[i]]
    685 if len(data[key]) == 0:
    686     data[key].num_nodes = node_ids[i].size(0)

IndexError: index is out of bounds for dimension with size 0
```
Any input would be greatly appreciated, thanks!

**Reply by rusty1s** [reply]
Do you have a small example to reproduce this issue? Any reason you need to go back in the first place?

**Reply by saiden89** [reply]
The two snippets above should be able to reproduce the issue; the `train` called in the second box is the same as defined in the upper one.

My (possibly invalid) reason for going back relates to the model's architecture that I use to predict links. Currently, it is implemented with a 3-module system: an encoder, an aggregator and a decoder/predictor. The encoder is a PyTorch `ModuleDict` with separate encoders for each edge type. The aggregator merges the individual embeddings together and produces a single embedding. Finally, the encoder/predictor is also a `ModuleDict` with separate models for each edge type. So my reason for going back would be to keep the separate "specialized" decoders instead of using a symmetric decoder *i.e.* inner product. I am aware that this may not be the best solution to the problem, so I'm always open to new ideas 😄.

**Reply by rusty1s** [reply]
Indeed, that code snippet helped me. I fixed it in https://github.com/pyg-team/pytorch_geometric/pull/6522.

**Reply by saiden89** [reply]
The fix works for the train data (*i.e.* when message passing and supervision edges are the some) but fails for the validation and test datasets. 
```python3
val.to_heterogeneous()
---------------------------------------------------------------------------
IndexError                                Traceback (most recent call last)
Cell In[4], line 18
     15 train, val, test = transform(data)
     16 print(val)
---> 18 test.to_heterogeneous()

File lib/python3.10/site-packages/torch_geometric/data/data.py:746, in Data.to_heterogeneous(self, node_type, edge_type, node_type_names, edge_type_names)
    744         elif isinstance(value, Tensor) and self.is_edge_attr(attr):
    745             cat_dim = self.__cat_dim__(attr, value)
--> 746             data[key][attr] = value.index_select(cat_dim, edge_ids[i])
    748 # Add global attributes.
    749 exclude_keys = set(data.keys) | {
    750     'node_type', 'edge_type', 'edge_index', 'num_nodes', 'ptr'
    751 }

IndexError: index out of range in self
```

**Reply by rusty1s** [reply]
This is now fixed in https://github.com/pyg-team/pytorch_geometric/pull/6529.

**Reply by saiden89** [reply]
Thanks for the fixes, however there are inconsistent results when `to_homogeneous` and `to_heterogeneous` are called in sequence on different data instances. To reproduce:
```python
train.to_homogeneous().to_heterogeneous() # works as expected

>>> HeteroData(
  node_label=[0],
  v0={
    x=[779, 77],
    y=[779]
  },
  (v0, e0, v0)={
    edge_index=[2, 5422],
    edge_weight=[5422]
  },
  (v0, e1, v0)={
    edge_index=[2, 5419],
    edge_weight=[5419]
  },
  (v0, e2, v0)={
    edge_index=[2, 5425],
    edge_weight=[5425]
  },
  (v0, e3, v0)={
    edge_index=[2, 5426],
    edge_weight=[5426]
  }
)
```
```python
val.to_homogeneous().to_heterogeneous() # edge_weight separated from edge type.

>>> HeteroData(
  node_label=[0],
  edge_weight=[4198],
  v0={
    x=[1055, 67],
    y=[1055]
  },
  (v0, e0, v0)={ edge_index=[2, 7357] },
  (v0, e1, v0)={ edge_index=[2, 7349] },
  (v0, e2, v0)={ edge_index=[2, 7351] },
  (v0, e3, v0)={ edge_index=[2, 7350] }
)
```

**Reply by rusty1s** [reply]
Yes, I think this is to be expected, since the shape of `edge_weight` and `edge_index` no longer match.

**Reply by saiden89** [reply]
Would you know how to retrieve the `edge_weight` for the different edges? If I happen to use a different decoder for each edge type I would need to know the edges-decoder mapping.

**Reply by rusty1s** [reply]
Wouldn't you still be able to use `edge_weight` and `edge_weight_index` since they are basically untouched?

**Reply by saiden89** [reply]
For the `train` set yes, the transformation is reversible and produces the same heterogeneous object back. But for the validation and test sets, only the `edge_index` gets mapped back to each specific `edge_type`. Indeed all the `edge_weight` get merged together with no index to map the back to the edge types they belong to. Besides the impossibility to use edge-specific decoders, it is also impossible to produce metrics on a per-edge basis. 
I'm not sure what you refer to with "`edge_weight` and `edge_weight_index` are basically untouched", but maybe I am missing the point completely.
Thank you for your help!

**Reply by rusty1s** [reply]
The problem with this approach is that the `edge_type` mapping is lost for validation and test since they have a different shape. Why don't you just do:
```python
data = FakeHeteroDataset(num_graphs=1, num_node_types=1, num_edge_types=4,
                         edge_dim=1, node_label=0)

transform = RandomLinkSplit(
    key="edge_weight",
    add_negative_train_samples=False,
    neg_sampling_ratio=0,
    edge_types=data[0].edge_types,
)

train, val, test = transform(data[0])
print(val)

**Reply by saiden89** [reply]
Because the original point of converting into a "batched" homogeneous network was to enable the creation of one dataloader per train, validation, and test split which include all the edge types, which is then passed to the Lightning model. To my understanding, since the model has different decoders and (possibly) decoders per edge type, it should be able to reconstruct the "original" heterogeneous graph before training. This currently works for the training set but not for the validation and test sets. 
I am summarizing the features here as they are spread across several comments:

- Goal: implement edge mini batching
- Data is heterogeneous with 4 edge types and 1 node type
- The objective is edge regression. Edges have associated probabilistic weights between 0 and 1.
- Model is a Lightning model because of distributed training
- Model has edge-specific encoders, an aggregator, and edge-specific decoders

If helpful I can share my progress so far with the model. I would also be happy to contribute with an example to help people with similar settings once the matter is settled.
Thank you again for your time and I apologize if my problem was unclear from the beginning.
